test/core/end2end/tests/max_connection_age.cc

```diff
@@ -157,68 +157,83 @@ static void test_max_age_forcibly_close(grpc_end2end_test_config config) {
-  op++;
-  op->op = GRPC_OP_SEND_STATUS_FROM_SERVER;
-  op->data.send_status_from_server.trailing_metadata_count = 0;
-  op->data.send_status_from_server.status = GRPC_STATUS_UNIMPLEMENTED;
-  grpc_slice status_details = grpc_slice_from_static_string("xyz");
-  op->data.send_status_from_server.status_details = &status_details;
-  op->flags = 0;
-  op->reserved = nullptr;
-  op++;
-  op->op = GRPC_OP_RECV_CLOSE_ON_SERVER;
-  op->data.recv_close_on_server.cancelled = &was_cancelled;
-  op->flags = 0;
-  op->reserved = nullptr;
-  op++;
-  error = grpc_call_start_batch(s, ops, static_cast<size_t>(op - ops), tag(102),
-                                nullptr);
-  GPR_ASSERT(GRPC_CALL_OK == error);
-  cqv->Expect(tag(102), true);
-  cqv->Verify();
+  grpc_event ev = grpc_completion_queue_next(
+      f.cq, gpr_inf_future(GPR_CLOCK_MONOTONIC), nullptr);
+  GPR_ASSERT(ev.type == GRPC_OP_COMPLETE);
+  GPR_ASSERT(ev.tag == tag(1) || ev.tag == tag(101));
+
+  if (ev.tag == tag(101)) {
+    // Request got through to the server before connection timeout
+
+    // Wait for the channel to reach its max age
+    cqv->VerifyEmpty(
+        grpc_core::Duration::Seconds(CQ_MAX_CONNECTION_AGE_WAIT_TIME_S));
+
+    // After the channel reaches its max age, we still do nothing here. And wait
+    // for it to use up its max age grace period.
+    cqv->Expect(tag(1), true);
+    cqv->Verify();
+
+    gpr_timespec expect_shutdown_time = grpc_timeout_milliseconds_to_deadline(
+        static_cast<int>(MAX_CONNECTION_AGE_MS *
+                         MAX_CONNECTION_AGE_JITTER_MULTIPLIER) +
+        MAX_CONNECTION_AGE_GRACE_MS + IMMEDIATE_SHUTDOWN_GRACE_TIME_MS);
+
+    gpr_timespec channel_shutdown_time = gpr_now(GPR_CLOCK_MONOTONIC);
+    GPR_ASSERT(gpr_time_cmp(channel_shutdown_time, expect_shutdown_time) < 0);
+
+    memset(ops, 0, sizeof(ops));
+    op = ops;
+    op->op = GRPC_OP_SEND_INITIAL_METADATA;
+    op->data.send_initial_metadata.count = 0;
+    op->flags = 0;
+    op->reserved = nullptr;
+    op++;
+    op->op = GRPC_OP_SEND_STATUS_FROM_SERVER;
+    op->data.send_status_from_server.trailing_metadata_count = 0;
+    op->data.send_status_from_server.status = GRPC_STATUS_UNIMPLEMENTED;
+    grpc_slice status_details = grpc_slice_from_static_string("xyz");
+    op->data.send_status_from_server.status_details = &status_details;
+    op->flags = 0;
+    op->reserved = nullptr;
+    op++;
+    op->op = GRPC_OP_RECV_CLOSE_ON_SERVER;
+    op->data.recv_close_on_server.cancelled = &was_cancelled;
+    op->flags = 0;
+    op->reserved = nullptr;
+    op++;
+    error = grpc_call_start_batch(s, ops, static_cast<size_t>(op - ops),
+                                  tag(102), nullptr);
+    GPR_ASSERT(GRPC_CALL_OK == error);
+    cqv->Expect(tag(102), true);
+    cqv->Verify();
+
+    GPR_ASSERT(0 == grpc_slice_str_cmp(call_details.method, "/foo"));
+    GPR_ASSERT(was_cancelled == 1);
+  } else {
+    // Request failed before getting to the server
```

Comment:
In this case, aren't we basically not testing what we're supposed to be testing here?

I'm wondering if it would be safer to just increase the max age time so that it's long enough to ensure that the call always gets to the server before it hits.  I realize that's more of a whack-a-mole solution, but it seems like it would at least give us some indication if things get slow enough that the test stops covering what it's supposed to.

Reply:
As per our offline conversation, it looks like this server-side code that's getting bypassed isn't really pertinent to the thing we're actually trying to test here, so it's okay to bypass it.